lib/forwarder/forwarder.go

```
@@ -0,0 +1,43 @@
+package forwarder
+
+import (
+	"context"
+	"io"
+	"sync"
+	"tcplb/lib/errors"
+)
+
+// MediocreForwarder is a implementation of the Forward operation.
+// This is a placeholder implementation that lacks robustness.
+type MediocreForwarder struct{}
+
+func (f MediocreForwarder) Forward(ctx context.Context, clientConn, upstreamConn DuplexConn) error {
+	// Caller is responsible for closing both DuplexConns, not us.
+	out := make(chan error, 2)
+	wg := sync.WaitGroup{}
+
+	copy := func(dst, src DuplexConn, out chan<- error) {
+		defer wg.Done()
+		// TODO FIXME add an idle timeout here that detects if neither
+		// of the two directions of copying have made any progress in
+		// some time window.
+		// TODO FIXME also honour cancellation by ctx
+		_, err := io.Copy(dst, src)
+		_ = dst.CloseWrite() // Inform peer at dst end that we're done writing.
```

Avoid swallowing errors, let's at least log.